--- nc/netcat.c.orig	2026-03-09 00:00:00.000000000 +0100
+++ nc/netcat.c	2026-03-09 00:00:01.000000000 +0100
@@ -226,7 +226,7 @@
       if (h_errno > 4)		/* oh no you don't, either */
 	fprintf (stderr, "preposterous h_errno: %d", h_errno);
       else
-	fprintf (stderr, h_errs[h_errno]);	/* handle it here */
+	fprintf (stderr, "%s", h_errs[h_errno]);	/* handle it here */
       h_errno = 0;				/* and reset for next call */
     }
 #endif
